// isotropic-error/lib/error.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});

var _getOwnPropertyDescriptor = require('babel-runtime/core-js/reflect/get-own-property-descriptor');

var _getOwnPropertyDescriptor2 = _interopRequireDefault(_getOwnPropertyDescriptor);

var _defineProperties = require('babel-runtime/core-js/object/define-properties');

var _defineProperties2 = _interopRequireDefault(_defineProperties);

var _stringify = require('babel-runtime/core-js/json/stringify');

var _stringify2 = _interopRequireDefault(_stringify);

var _isotropicMake = require('isotropic-make');

var _isotropicMake2 = _interopRequireDefault(_isotropicMake);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var _Error = (0, _isotropicMake2.default)(Error, {
    toString: function toString() {
        var string = this.name || 'Error';

        if (this.message) {
            string += ': ' + this.message;
        }

        return string;
    },
    _getMessage: function _getMessage() {
        return this._message || this.error && this.error.message || void null;
    },
    _getName: function _getName() {
        return this._name || this.error && this.error.name || void null;
    },
    _getStack: function _getStack() {
        if (!this._stack) {
            this._stack = this.toString();

            if (this.details) {
                this._stack += '\nDetails: ' + (0, _stringify2.default)(this.details, null, 4);
            }

            var prepareStackTrace = Error.prepareStackTrace;

            Error.prepareStackTrace = _Error._prepareStackTrace;
            this._stack += '\n' + this._nativeGetStack();
            Error.prepareStackTrace = prepareStackTrace;

            if (this.error) {
                this._stack += '\n-> ' + (this.error.stack || this.error);
            }
        }

        return this._stack;
    },
    _init: function _init() {
        var _ref = arguments.length <= 0 || arguments[0] === undefined ? {} : arguments[0];

        var details = _ref.details;
        var error = _ref.error;
        var message = _ref.message;
        var name = _ref.name;

        var nativeError = void 0;

        if (Error.captureStackTrace) {
            Error.captureStackTrace(this, _Error);
        } else {
            try {
                throw new Error();
            } catch (error) {
                nativeError = error;
            }
        }

        (0, _defineProperties2.default)(this, {
            details: {
                enumerable: true,
                value: details
            },
            error: {
                enumerable: true,
                value: error
            },
            message: {
                enumerable: true,
                get: this._getMessage
            },
            name: {
                enumerable: true,
                get: this._getName
            },
            stack: {
                get: this._getStack
            },
            _message: {
                value: message
            },
            _name: {
                value: name
            },
            _nativeGetStack: {
                value: nativeError ? function () {
                    return nativeError.stack;
                } : (0, _getOwnPropertyDescriptor2.default)(this, 'stack').get
            },
            _stack: {
                writable: true
            }
        });

        return this;
    }
}, {
    _prepareStackTrace: function _prepareStackTrace(errorObject, structuredStackTrace) {
        return structuredStackTrace.map(function (stackFrame) {
            var line = void 0;

            try {
                line = stackFrame.toString();
            } catch (error0) {
                try {
                    line = '<error: ' + error0 + '>';
                } catch (error1) {
                    line = '<error>';
                }
            }

            return '    at ' + line;
        }).join('\n');
    }
});

exports.default = _Error;
